File: android_mobile/src/main/java/com/alexstyl/specialdates/settings/DailyReminderFragment.java
package com.alexstyl.specialdates.settings;

import android.Manifest;
import android.content.Context;
import android.content.pm.PackageManager;
import android.media.Ringtone;
import android.media.RingtoneManager;
import android.net.Uri;
import android.os.Bundle;
import android.os.Vibrator;
import android.preference.CheckBoxPreference;
import android.preference.Preference;
import android.preference.Preference.OnPreferenceChangeListener;
import android.support.annotation.NonNull;
import android.text.format.DateFormat;

import com.alexstyl.android.AlarmManagerCompat;
import com.alexstyl.android.preferences.widget.TimePreference;
import com.alexstyl.specialdates.AppComponent;
import com.alexstyl.specialdates.MementoApplication;
import com.alexstyl.specialdates.R;
import com.alexstyl.specialdates.TimeOfDay;
import com.alexstyl.specialdates.analytics.Analytics;
import com.alexstyl.specialdates.dailyreminder.DailyReminderPreferences;
import com.alexstyl.specialdates.dailyreminder.DailyReminderScheduler;
import com.alexstyl.specialdates.permissions.PermissionChecker;
import com.alexstyl.specialdates.ui.base.MementoPreferenceFragment;

import javax.inject.Inject;
import java.util.Calendar;

public class DailyReminderFragment extends MementoPreferenceFragment {

    private static final int EXTERNAL_STORAGE_REQUEST_CODE = 15;

    private CheckBoxPreference enablePreference;
    private RingtonePreference ringtonePreference;
    private TimePreference timePreference;
    private PermissionChecker permissionChecker;
    private DailyReminderScheduler scheduler;
    private DailyReminderPreferences preferences;
    @Inject Analytics analytics;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Context context = getActivity().getApplicationContext();
        scheduler = new DailyReminderScheduler(AlarmManagerCompat.from(context), context);
        AppComponent applicationModule = ((MementoApplication) getActivity().getApplication()).getApplicationModule();
        applicationModule.inject(this);
        addPreferencesFromResource(R.xml.preference_dailyreminder);
        permissionChecker = new PermissionChecker(getActivity());
        enablePreference = findPreference(R.string.key_daily_reminder);

        preferences = DailyReminderPreferences.newInstance(getActivity());
        enablePreference.setOnPreferenceChangeListener(new OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object newValue) {
                boolean isEnabled = (boolean) newValue;
                preferences.setEnabled(isEnabled);

                if (isEnabled) {
                    analytics.trackDailyReminderEnabled();
                    scheduler.setupReminder(preferences);
                } else {
                    analytics.trackDailyReminderDisabled();
                    scheduler.cancelReminder();
                }
                return true;
            }

        });

        ringtonePreference = findPreference(R.string.key_daily_reminder_ringtone);
        ringtonePreference.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                if (permissionChecker.canReadExternalStorage()) {
                    // the permission exists. Let the system handle the event
                    return false;
                } else {
                    requestPermissions(new String[]{Manifest.permission.READ_EXTERNAL_STORAGE}, EXTERNAL_STORAGE_REQUEST_CODE);
                    return true;
                }
            }

        });
        ringtonePreference.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {

            @Override
            public boolean onPreferenceChange(Preference preference, Object newValue) {
                updateRingtoneSummaryFor(Uri.parse((String) newValue));
                return true;
            }
        });

        timePreference = findPreference(R.string.key_daily_reminder_time);
        timePreference.setOnPreferenceChangeListener(new OnPreferenceChangeListener() {

            @Override
            public boolean onPreferenceChange(Preference preference, Object newValue) {
                int[] time = (int[]) newValue;
                TimeOfDay timeOfDay = new TimeOfDay(time[0], time[1]);
                updateTimeSet(timeOfDay);
                analytics.trackDailyReminderTimeUpdated(timeOfDay);
                preferences.setDailyReminderTime(timeOfDay);
                scheduler.updateReminderTime(timeOfDay);
                return true;
            }
        });

        hideVibratorSettingIfNotPresent();
    }

    private void hideVibratorSettingIfNotPresent() {
        if (hasNoVibratorHardware()) {
            Preference vibratePreference = findPreference(getString(R.string.key_daily_reminder_vibrate_enabled));
            getPreferenceScreen().removePreference(vibratePreference);
        }
    }

    private boolean hasNoVibratorHardware() {
        Vibrator vibrator = (Vibrator) getActivity().getSystemService(Context.VIBRATOR_SERVICE);
        return !vibrator.hasVibrator();
    }

    @Override
    public void onResume() {
        super.onResume();
        enablePreference.setChecked(preferences.isEnabled());
        updateRingtoneSummaryFor(preferences.getRingtoneSelected());

        TimeOfDay timeOfDay = preferences.getDailyReminderTimeSet();
        updateTimeSet(timeOfDay);
    }

    private void updateTimeSet(TimeOfDay time) {
        String timeString = getStringHour(time);
        String summary = String.format(getString(R.string.daily_reminder_time_summary), timeString);
        timePreference.setSummary(summary);

    }

    private String getStringHour(TimeOfDay time) {
        Calendar cal = Calendar.getInstance();
        cal.set(Calendar.HOUR_OF_DAY, time.getHours());
        cal.set(Calendar.MINUTE, time.getMinutes());
        return getHour(getActivity(), cal).toString();
    }

    // Char sequence for a 12 hour format.
    private static final CharSequence DEFAULT_FORMAT_12_HOUR = "hh:mm a";
    // Char sequence for a 24 hour format.
    private static final CharSequence DEFAULT_FORMAT_24_HOUR = "kk:mm";

    public static CharSequence getHour(Context context, Calendar cal) {
        boolean is24Hour = DateFormat.is24HourFormat(context);
        if (is24Hour) {
            return DateFormat.format(DEFAULT_FORMAT_24_HOUR, cal);
        } else {
            return DateFormat.format(DEFAULT_FORMAT_12_HOUR, cal);
        }

    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == EXTERNAL_STORAGE_REQUEST_CODE && grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            ringtonePreference.onClick();
        }
    }

    private void updateRingtoneSummaryFor(Uri ringtoneUri) {
        String name = null;
        if (ringtoneUri.toString().length() > 0) {
            Ringtone ringtone = RingtoneManager.getRingtone(getActivity(), ringtoneUri);
            if (ringtone != null) {
                name = ringtone.getTitle(getActivity());
            }
        } else {
            name = getString(R.string.no_sound);
        }
        ringtonePreference.setSummary(name);
    }
}

File: android_mobile/src/main/java/com/alexstyl/specialdates/settings/DailyReminderFragment.kt
package com.alexstyl.specialdates.settings

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.pm.PackageManager
import android.media.RingtoneManager
import android.net.Uri
import android.os.Bundle
import android.os.Vibrator
import android.preference.CheckBoxPreference
import android.preference.Preference
import android.preference.Preference.OnPreferenceChangeListener
import android.preference.RingtonePreference
import android.text.format.DateFormat
import com.alexstyl.android.preferences.widget.TimePreference
import com.alexstyl.specialdates.CrashAndErrorTracker
import com.alexstyl.specialdates.MementoApplication
import com.alexstyl.specialdates.R
import com.alexstyl.specialdates.TimeOfDay
import com.alexstyl.specialdates.analytics.Analytics
import com.alexstyl.specialdates.dailyreminder.DailyReminderOreoChannelCreator
import com.alexstyl.specialdates.dailyreminder.DailyReminderScheduler
import com.alexstyl.specialdates.dailyreminder.DailyReminderUserSettings
import com.alexstyl.specialdates.permissions.MementoPermissions
import com.alexstyl.specialdates.ui.base.MementoPreferenceFragment
import java.net.URI
import java.util.Calendar
import javax.inject.Inject

class DailyReminderFragment : MementoPreferenceFragment() {

    private lateinit var enablePreference: CheckBoxPreference
    private lateinit var timePreference: TimePreference

    private var ringtoneLegacyPreference: ClickableRingtonePreference? = null

    lateinit var permissions: MementoPermissions
        @Inject set
    lateinit var dailyReminderScheduler: DailyReminderScheduler
        @Inject set
    lateinit var analytics: Analytics
        @Inject set
    lateinit var tracker: CrashAndErrorTracker
        @Inject set
    lateinit var preferences: DailyReminderUserSettings
        @Inject set
    lateinit var navigator: DailyReminderNavigator
        @Inject set
    lateinit var channelCreator: DailyReminderOreoChannelCreator
        @Inject set


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val applicationModule = (activity!!.application as MementoApplication).applicationModule
        applicationModule.inject(this)

        addPreferencesFromResource(R.xml.preference_dailyreminder)

        enablePreference = findPreferenceOrThrow(R.string.key_daily_reminder)
        enablePreference.onPreferenceChangeListener = OnPreferenceChangeListener { _, newValue ->
            val isEnabled = newValue as Boolean
            preferences.setEnabled(isEnabled)

            if (isEnabled) {
                analytics.trackDailyReminderEnabled()
                dailyReminderScheduler.scheduleReminderFor(preferences.getTimeSet())
            } else {
                analytics.trackDailyReminderDisabled()
                dailyReminderScheduler.cancelReminder()
            }
            true
        }

        timePreference = findPreferenceOrThrow(R.string.key_daily_reminder_time)
        timePreference.onPreferenceChangeListener = OnPreferenceChangeListener { _, newValue ->
            val time = newValue as IntArray
            val timeOfDay = TimeOfDay(time[0], time[1])
            updateTimeSet(timeOfDay)
            analytics.trackDailyReminderTimeUpdated(timeOfDay)
            preferences.setDailyReminderTime(timeOfDay)
            dailyReminderScheduler.scheduleReminderFor(timeOfDay)
            true
        }

        ringtoneLegacyPreference = findPreference(R.string.key_daily_reminder_ringtone)
        ringtoneLegacyPreference?.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            if (permissions.canReadExternalStorage()) {
                // the permission exists. Let the system handle the event
                false
            } else {
                requestPermissions(arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE), EXTERNAL_STORAGE_REQUEST_CODE)
                true
            }
        }
        ringtoneLegacyPreference?.onPreferenceChangeListener = OnPreferenceChangeListener { _, newValue ->
            val ringtoneUri = URI.create(newValue as String)
            ringtoneLegacyPreference?.updateRingtoneSummaryWith(ringtoneUri)
            true
        }

        val setVibrationPreference = findPreference<Preference>(R.string.key_daily_reminder_vibrate_enabled)
        setVibrationPreference?.apply {
            if (hasNoVibratorHardware()) {
                preferenceScreen.removePreference(this)
            }
        }
        setVibrationPreference?.onPreferenceChangeListener =
                OnPreferenceChangeListener { _: Preference, _: Any ->
                    true
                }

        findPreference<Preference>(R.string.key_daily_reminder_advanced_settings)?.setOnPreferenceClickListener { _ ->
            navigator.openAdvancedSettings(activity as Activity)
            true
        }

    }

    private fun hasNoVibratorHardware(): Boolean {
        val vibrator = activity!!.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
        return !vibrator.hasVibrator()
    }

    override fun onResume() {
        super.onResume()
        enablePreference.isChecked = preferences.isEnabled()
        ringtoneLegacyPreference?.updateRingtoneSummaryWith(preferences.getRingtone())

        val timeOfDay = preferences.getTimeSet()
        updateTimeSet(timeOfDay)
    }


    private fun updateTimeSet(time: TimeOfDay) {
        val timeString = getStringHour(time)
        val summary = String.format(getString(R.string.daily_reminder_time_summary), timeString)
        timePreference.summary = summary

    }

    private fun getStringHour(time: TimeOfDay): String {
        val cal = Calendar.getInstance()
        cal.set(Calendar.HOUR_OF_DAY, time.hours)
        cal.set(Calendar.MINUTE, time.minutes)
        return getHour(activity, cal).toString()
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == EXTERNAL_STORAGE_REQUEST_CODE && grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            ringtoneLegacyPreference!!.onClick()
        }
    }

    private fun RingtonePreference.updateRingtoneSummaryWith(ringtoneUri: URI) {
        var name: String? = null
        if (ringtoneUri.toString().isNotEmpty()) {
            val ringtone = RingtoneManager.getRingtone(activity, Uri.parse(ringtoneUri.toString()))
            if (ringtone != null) {
                name = ringtone.getTitle(activity)
            }
        } else {
            name = getString(R.string.no_sound)
        }
        summary = name
    }

    companion object {

        private const val EXTERNAL_STORAGE_REQUEST_CODE = 15

        // Char sequence for a 12 hour format.
        private const val DEFAULT_FORMAT_12_HOUR = "hh:mm a"
        // Char sequence for a 24 hour format.
        private const val DEFAULT_FORMAT_24_HOUR = "kk:mm"

        fun getHour(context: Context?, cal: Calendar): CharSequence {
            val is24Hour = DateFormat.is24HourFormat(context)
            return if (is24Hour) {
                DateFormat.format(DEFAULT_FORMAT_24_HOUR, cal)
            } else {
                DateFormat.format(DEFAULT_FORMAT_12_HOUR, cal)
            }

        }
    }
}

